File: baselibs/java/openhpi_baselib/HpiSession.java
package org.openhpi;

import java.util.Collection;
import java.util.concurrent.atomic.AtomicLong;
import java.util.LinkedList;


class HpiSession
{
    private static AtomicLong next_local_sid = new AtomicLong( 1L );


    private long                   local_sid;
    private long                   remote_sid;
    private HpiDomain              domain;
    private LinkedList<HpiMarshal> marshals;


    public HpiSession( HpiDomain domain )
    {
        this.local_sid  = next_local_sid.getAndIncrement();
        this.remote_sid = 0;
        this.domain     = new HpiDomain( domain );
        this.marshals   = new LinkedList<HpiMarshal>();
    }

    public synchronized void close()
    {
        while( !marshals.isEmpty() ) {
            HpiMarshal m = marshals.removeFirst();
            m.close();
        }
    }

    public long getLocalSid()
    {
        return local_sid;
    }

    public long getRemoteSid()
    {
        return remote_sid;
    }

    public void setRemoteSid( long remote_sid )
    {
        this.remote_sid = remote_sid;
    }

    public long getRemoteDid()
    {
        return domain.getRemoteDid();
    }

    public synchronized HpiMarshal getMarshal()
    {
        HpiMarshal m;
        if ( !marshals.isEmpty() ) {
            m = marshals.removeFirst();
        } else {
            m = new HpiMarshal();
            boolean rc = m.open( domain.getRemoteHost(), domain.getRemotePort() );
            if ( !rc ) {
                m = null;
            }
        }
        if ( m != null ) {
            m.reset();
        }
        return m;
    }

    public synchronized void putMarshal( HpiMarshal m )
    {
        m.reset();
        marshals.addFirst( m );
    }
};

File: baselibs/csharp/openhpi_baselib/HpiSession.cs
using System;
using System.Collections.Generic;
using System.Threading;


namespace org {
namespace openhpi {


internal class HpiSession
{
    private static long cnt = 0;


    private long                    local_sid;
    private long                    remote_sid;
    private HpiDomain               domain;
    private LinkedList<OhpiMarshal> marshals;


    public HpiSession( HpiDomain domain )
    {
        // TODO investigate overflow issue
        this.local_sid  = Interlocked.Increment( ref cnt );
        this.remote_sid = 0;
        this.domain     = new HpiDomain( domain );
        this.marshals   = new LinkedList<OhpiMarshal>();
    }

    public void Close()
    {
        lock ( marshals )
        {
            while( marshals.Count > 0 ) {
                OhpiMarshal m = marshals.First.Value;
                marshals.RemoveFirst();
                m.Close();
            }
        }
    }

    public long GetLocalSid()
    {
        return local_sid;
    }

    public long GetRemoteSid()
    {
        return remote_sid;
    }

    public void SetRemoteSid( long remote_sid )
    {
        this.remote_sid = remote_sid;
    }

    public long GetRemoteDid()
    {
        return domain.GetRemoteDid();
    }

    public OhpiMarshal GetMarshal()
    {
        OhpiMarshal m;

        lock ( marshals )
        {
            if ( marshals.Count > 0 ) {
                m = marshals.First.Value;
                marshals.RemoveFirst();
            } else {
                m = new OhpiMarshal();
                bool rc = m.Open( domain.GetRemoteHost(), domain.GetRemotePort() );
                if ( !rc ) {
                    m = null;
                }
            }
            if ( m != null ) {
                m.Reset();
            }
        }

        return m;
    }

    public void PutMarshal( OhpiMarshal m )
    {
        m.Reset();
        lock ( marshals )
        {
            marshals.AddFirst( m );
        }
    }
};


}; // namespace openhpi
}; // namespace org

